jcl/src/java.base/share/classes/openj9/internal/foreign/abi/UpcallMHMetaData.java

```diff
@@ -111,48 +116,61 @@ final class UpcallMHMetaData {
 		/*[ENDIF] JAVA_SPEC_VERSION >= 21 */
 	}
 
-	/* Determine whether the memory segment of the passed-in/returned pointer is allocated
-	 * in the native memory or not and return its native address if valid.
+	/*[IF JAVA_SPEC_VERSION >= 21]*/
+	/* Determine whether or not the native address of the passed-in/returned pointer is aligned against
+	 * its target layout's alignment and return its segment object if valid.
 	 *
 	 * Note:
-	 * The method is shared in java (downcall) and in native (upcall) via the calling-in from the dispatcher.
+	 * The method is shared in downcall and upcall.
 	 */
-	static long getNativeArgRetSegmentOfPtr(MemorySegment argRetSegmentOfPtr) {
-		if (!argRetSegmentOfPtr.isNative()) {
-			throw new IllegalArgumentException("Heap segment not allowed: " + argRetSegmentOfPtr);
+	static MemorySegment getArgRetAlignedSegmentOfPtr(long addrValue, MemoryLayout layout) {
+		AddressLayout addrLayout = (AddressLayout)layout;
+		if (!addrLayout.targetLayout().isEmpty()
```

Comment:
Please extract `addrLayout.targetLayout()` to a local variable.

Reply:
Fixed.